cargo-miri/bin.rs

```diff
@@ -602,36 +604,28 @@ fn phase_cargo_rustc(args: env::Args) {
         _ => {},
     }
 
-    if !print && is_runnable_crate() {
-        macro_rules! store {
-            ($info:expr) => {
-                let info = $info;
-                let filename = out_filename("", "");
-                if verbose {
-                    eprintln!("[cargo-miri rustc] writing run info to `{}`", filename.display());
-                }
-                info.store(&filename);
-                // For Windows, do the same thing again with `.exe` appended to the filename.
-                // (Need to do this here as cargo moves that "binary" to a different place before
-                // running it.)
-                info.store(&out_filename("", ".exe"));
-                return;
-            };
-        }
-        if target_crate {
-            // This is the binary or test crate that we want to interpret under Miri.
-            // But we cannot run it here, as cargo invoked us as a compiler -- our
-            // stdin and stdout are not like we want them.
-            // Instead of compiling, we write JSON into the output file with all the
-            // relevant command-line flags and environment variables; this is used when
-            // cargo calls us again in the CARGO_TARGET_RUNNER phase.
-            store!(CrateRunInfo::collect(args));
-        } else if ArgFlagValueIter::new("--extern").any(|krate| krate == "proc_macro") {
-            // This is a "runnable" `proc-macro` crate (unit tests). We do not support
-            // interpreting that under Miri now, so we write a JSON file to (display a
-            // helpful message and) skip it in the runner phase.
-            store!(CrateRunInfo::Skip);
+    let store_json = |info: CrateRunInfo| {
+        let filename = out_filename("", "");
+        if verbose {
+            eprintln!("[cargo-miri rustc] writing run info to `{}`", filename.display());
         }
+        info.store(&filename);
+        // For Windows, do the same thing again with `.exe` appended to the filename.
+        // (Need to do this here as cargo moves that "binary" to a different place before running it.)
+        info.store(&out_filename("", ".exe"));
+    };
+
+    if !print && target_crate && is_runnable_crate() {
+        store_json(CrateRunInfo::collect(args));
+        return;
+    }
+
+    if !print
```

Comment:
`--extern` existing implies `!print`, doesn't it?

Reply:
rustc happily accepts `--extern proc_macro --print=cfg`. I'm not sure would Cargo do that, though.

Reply:
Hm good point. Now I see why you made this one `if`.

We should also avoid calling `is_runnable_crate()` twice. So what about something like this?
```rust
let run_crate = !print && is_runnable_crate();
```

Reply:
> We should also avoid calling `is_runnable_crate()` twice. So what about something like this?

That's fair, but I'm nore sure if `run_crate` is a good name, as unit tests of `proc-macro` crates are skipped.
I can't think of a good name now, but I think changing the condition to `!print && !target_crate && is_runnable_crate() && ArgFlagValueIter::new("--extern").any(|krate| krate == "proc_macro")` could avoid another call to `is_runnable_crate()`. That's more complex, though.

Reply:
> but I'm not sure if run_crate is a good name, as unit tests of proc-macro crates are skipped.

Yeah... so, `runnable_crate`?

Reply:
> Yeah... so, `runnable_crate`?

There's also `!print`. I don't know can `runnable_crate` covers that?

Reply:
Well, `print` is really a separate case. It means there's no crate to begin with, just a compiler query.